### DISCUSSION TITLE
Custom BIRT plugins seem not to be loaded on newer Debian and Ubuntu server systems

### DISCUSSION THREAD

**Discussion by blatra** [eclipse-birt/birt#994] [General]
I've written a small plugin for BIRT that provide some functions I can use in scripts in BIRT reports.
They run fine (using BIRT OSGi Runtime 4.6 and also 4.8 version on an AdoptOpenJDK jdk8u252-b09) on Ubuntu server 16.04 and 18.04 as well as on a Debian 10 desktop system.

But unfortunally I can not get them to run on Debian 10 & 11 server system as well on ubuntu server 20.04.
The report crashes and I see errors like this:
`Error 2: There are errors evaluating script "this.text = MyTestPlugin.testFunction();":
ReferenceError: "MyTestPlugin" is not defined. (/report/body/label[@id="6"]/method[@name="onCreate"]#1)`

I've checked the system configurations, file ownerships and permissions over and over, but did not find the cause for this error.
It seems like the plugin does not get loaded correctly.

I wonder if others had similar problems and were able to fix them.
Maybe there is a dependency of OSGi or BIRT that the operating system has to provide?

**Comment by wimjongman** [comment]
Can you post the manifest of your plugin?
Are you using a jre vs a jdk on those other systems?
Are the java versions compatible?

**Comment by blatra** [comment]
Yes, here's the manifest:

> Manifest-Version: 1.0
Bundle-ManifestVersion: 2
Bundle-Name: %Bundle-Name
Bundle-SymbolicName: script.test.core;singleton:=true
Bundle-Version: 1.0.0.qualifier
Bundle-Activator: script.test.core.Activator
Bundle-RequiredExecutionEnvironment: JavaSE-1.8
Bundle-Vendor: %Bundle-Vendor
Bundle-ActivationPolicy: lazy
Require-Bundle: org.eclipse.birt.report.engine

I'm using the same JDK on every system.

I also did some research. On the systems with BIRT 4.6 where it does not work I have run a report with following Script in the onCreate of a label:
`this.text = org.eclipse.birt.core.framework.Platform.getExtensionRegistry();`
And I get these errors:

> WARNUNG: There are errors evaluating script "this.text = org.eclipse.birt.core.framework.Platform.getExtensionRegistry();":
org/eclipse/core/runtime/IAdapterManager

and

> Caused by: java.lang.NoClassDefFoundError: org/eclipse/core/runtime/IAdapterManager
        at java.lang.Class.getDeclaredMethods0(Native Method)
        at java.lang.Class.privateGetDeclaredMethods(Class.java:2701)
        at java.lang.Class.privateGetPublicMethods(Class.java:2902)
        at java.lang.Class.getMethods(Class.java:1615)
        at org.mozilla.javascript.JavaMembers.discoverAccessibleMethods(JavaMembers.java:346)
        at org.mozilla.javascript.JavaMembers.discoverAccessibleMethods(JavaMembers.java:302)
        at org.mozilla.javascript.JavaMembers.reflect(JavaMembers.java:419)
        at org.mozilla.javascript.JavaMembers.<init>(JavaMembers.java:43)
        at org.mozilla.javascript.JavaMembers.lookupClass(JavaMembers.java:807)
        at org.mozilla.javascript.NativeJavaClass.initMembers(NativeJavaClass.java:49)

But usually I should get a label with text like this:
> org.eclipse.birt.core.framework.eclipse.EclipseExtensionRegistry@18d95a08

**Comment by wimjongman** [comment]
this.text = org.eclipse.birt.core.framework.Platform.getExtensionRegistry();

Are you using the OSGi runtime? The extension registry is normally not available outside Eclipse. When running the OSGi runtime I don't think you have access to it.

Either way, the org.eclipse.core.runtime jar seems missing in this configuration.

**Comment by blatra** [comment]
Yes I'm using the OSGi runtime.
And yes, you are right. This report seems to only work inside of eclipse.

But, I can confirm that org.eclipse.core.runtime is present in my installation:
> root@debian:/opt/birt# find ./ -type f -name "org.eclipse.core.runtime*"
./ReportEngine/plugins/org.eclipse.core.runtime_3.12.0.v20160606-1342.jar

**Comment by blatra** [comment]
It seems the error happens because of the user account or the corresponding systemd service that are used for starting the java process.

After I assign a login shell to the system user like
`usermod -s /bin/bash birtuser`
I'm able to run the process via
`runuser -l birtuser -c "/opt/birtprocessor/startup.sh"`

Afterwards I can use my plugin in the reports without issues.

I have no idea how this can affect my plugin inside of the runtime, yet.
